sirepo/sim_api/jupyterhublogin.py

```diff
@@ -0,0 +1,166 @@
+# -*- coding: utf-8 -*-
+u"""API's for jupyterhublogin sim
+
+:copyright: Copyright (c) 2020 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from __future__ import absolute_import, division, print_function
+from pykern import pkconfig, pkio
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdp, pkdlog
+import flask
+import py.error
+import random
+import re
+import sirepo.api_perm
+import sirepo.auth
+import sirepo.auth_db
+import sirepo.events
+import sirepo.http_reply
+import sirepo.http_request
+import sirepo.srdb
+import sirepo.uri_router
+import sirepo.util
+import sqlalchemy
+import string
+
+cfg = None
+
+#: Used by auth_db. Sirepo record of each jupyterhub user.
+JupyterhubUser = None
+
+
+@sirepo.api_perm.require_user
+def api_migrateJupyterhub():
+    if not cfg.rs_jupyter_migrate:
+        sirepo.util.raise_forbidden('migrate not enabled')
+    d = sirepo.http_request.parse_json()
+    if not d.doMigration:
+        return sirepo.http_reply.gen_redirect('jupyterHub')
+    return sirepo.uri_router.call_api(
+        'authGithubLogin',
+        kwargs=PKDict(simulation_type='jupyterhublogin'),
+    )
+
+
+@sirepo.api_perm.require_user
+def api_redirectJupyterHub():
+    is_new_user = _create_user()
+    if not cfg.rs_jupyter_migrate or not is_new_user:
+        return sirepo.http_reply.gen_redirect('jupyterHub')
+    return sirepo.http_reply.gen_json_ok()
+
+
+def jupyterhub_user_name(have_simulation_db=True):
+    return _user_name(sirepo.auth.logged_in_user(check_path=have_simulation_db))
+
+
+def init_apis(*args, **kwargs):
+    global cfg
+
+    cfg = pkconfig.init(
+        dst_db_root=(
+            pkio.py_path(sirepo.srdb.root()).join('jupyterhub'),
+            pkio.py_path,
+            'new jupyter user db',
+        ),
+        rs_jupyter_migrate=(False, bool, 'give user option to migrate data from jupyter.radiasoft.org'),
+        src_db_root=(
+            pkio.py_path('/var/empty'),
+            pkio.py_path,
+            'existing jupyter user db (ex /srv/jupyterhub)',
+        ),
+        uri_root=('jupyter', str, 'the root uri of jupyterhub'),
+    )
+    sirepo.auth_db.init_model(_init_model)
+    sirepo.events.register({
+        'auth_logout': _event_auth_logout,
+        'end_api_call': _event_end_api_call,
+        'github_authorized': _event_github_authorized,
+    })
+
+
+def _create_user():
+    def __user_name(logged_in_user_name):
```

Comment:
having the word user_name here is confusing to me :) How about "name" or "value" or "auth_name"? The problem is that __user_name returns a user name, but it is really a juptyerhub user name (implicitly by the name of the module). Maybe if the method were called `__clean_user_name`.